rmw_fastrtps_shared_cpp/src/custom_publisher_info.cpp

```diff
@@ -21,156 +21,202 @@
+        publisher_info_->data_writer_->get_liveliness_lost_status(liveliness_lost_status_);
+        callback(user_data, liveliness_lost_status_.total_count_change);
+        liveliness_lost_status_.total_count_change = 0;
+        break;
+      case RMW_EVENT_OFFERED_DEADLINE_MISSED:
+        publisher_info_->data_writer_->get_offered_deadline_missed_status(
+          offered_deadline_missed_status_);
+        callback(
+          user_data,
+          offered_deadline_missed_status_.total_count_change);
+        offered_deadline_missed_status_.total_count_change = 0;
+        break;
+      case RMW_EVENT_OFFERED_QOS_INCOMPATIBLE:
+        publisher_info_->data_writer_->get_offered_incompatible_qos_status(
+          incompatible_qos_status_);
+        callback(
+          user_data,
+          incompatible_qos_status_.total_count_change);
+        incompatible_qos_status_.total_count_change = 0;
+        break;
+      default:
+        break;
+    }
+
+    user_data_[event_type] = user_data;
+    on_new_event_cb_[event_type] = callback;
+
+    eprosima::fastdds::dds::StatusMask status_mask =
+      publisher_info_->data_writer_->get_status_mask();
+    publisher_info_->data_writer_->set_listener(
+      this, status_mask << rmw_fastrtps_shared_cpp::internal::rmw_event_to_dds_statusmask(
+        event_type));
+  } else {
+    eprosima::fastdds::dds::StatusMask status_mask =
+      publisher_info_->data_writer_->get_status_mask();
+    publisher_info_->data_writer_->set_listener(
+      this, status_mask >> rmw_fastrtps_shared_cpp::internal::rmw_event_to_dds_statusmask(
+        event_type));
+
+    user_data_[event_type] = nullptr;
+    on_new_event_cb_[event_type] = nullptr;
+  }
+}
+
 void
 PubListener::on_offered_deadline_missed(
   eprosima::fastdds::dds::DataWriter * /* writer */,
   const eprosima::fastdds::dds::OfferedDeadlineMissedStatus & status)
 {
-  std::lock_guard<std::mutex> lock(internalMutex_);
-
-  // the change to liveliness_lost_count_ needs to be mutually exclusive with
-  // rmw_wait() which checks hasEvent() and decides if wait() needs to be called
-  ConditionalScopedLock clock(conditionMutex_, conditionVariable_);
+  std::unique_lock<std::mutex> lock_mutex(on_new_event_m_);
 
   // Assign absolute values
   offered_deadline_missed_status_.total_count = status.total_count;
   // Accumulate deltas
   offered_deadline_missed_status_.total_count_change += status.total_count_change;
 
-  deadline_changes_.store(true, std::memory_order_relaxed);
-
-  std::unique_lock<std::mutex> lock_mutex(on_new_event_m_);
+  deadline_changes_ = true;
 
-  if (on_new_event_cb_) {
-    on_new_event_cb_(user_data_, 1);
-  } else {
-    unread_events_count_++;
+  if (on_new_event_cb_[RMW_EVENT_OFFERED_DEADLINE_MISSED]) {
+    on_new_event_cb_[RMW_EVENT_OFFERED_DEADLINE_MISSED](user_data_[RMW_EVENT_OFFERED_DEADLINE_MISSED
+      ], 1);
   }
+
+  event_guard[RMW_EVENT_OFFERED_DEADLINE_MISSED].set_trigger_value(true);
```

Comment:
This code is repeated several times with different event types.
I think it is worth having a `trigger_event(rmw_event_type_t event_type)` method.

Reply:
Done in 68aacaa